File: app/src/main/java/mega/privacy/android/app/AndroidCompletedTransfer.java
package mega.privacy.android.app;


import android.os.Parcel;
import android.os.Parcelable;

import java.io.File;

import mega.privacy.android.app.objects.SDTransfer;
import nz.mega.sdk.MegaError;
import nz.mega.sdk.MegaNode;
import nz.mega.sdk.MegaTransfer;

import static mega.privacy.android.app.utils.MegaNodeUtil.*;
import static mega.privacy.android.app.utils.OfflineUtils.*;
import static mega.privacy.android.app.utils.SDCardUtils.getSDCardTargetPath;
import static mega.privacy.android.app.utils.StringResourcesUtils.*;
import static mega.privacy.android.app.utils.TextUtil.*;
import static mega.privacy.android.app.utils.Util.*;
import static nz.mega.sdk.MegaApiJava.INVALID_HANDLE;

public class AndroidCompletedTransfer implements Parcelable {

    private long id;
    private String fileName;
    private int type;
    private int state;
    private String size;
    private String nodeHandle;
    private String path;
    private boolean isOfflineFile;
    private long timeStamp;
    private String error;
    private String originalPath;
    private long parentHandle;

    public AndroidCompletedTransfer(long id, String fileName, int type, int state, String size,
                                    String nodeHandle, String path, boolean isOfflineFile,
                                    long timeStamp, String error, String originalPath,
                                    long parentHandle) {
        this.id = id;
        this.fileName = fileName;
        this.type = type;
        this.state = state;
        this.size = size;
        this.nodeHandle = nodeHandle;
        this.path = removeLastFileSeparator(path);
        this.isOfflineFile = isOfflineFile;
        this.timeStamp = timeStamp;
        this.error = error;
        this.originalPath = originalPath;
        this.parentHandle = parentHandle;
    }

    public AndroidCompletedTransfer (MegaTransfer transfer, MegaError error) {
        this.fileName = transfer.getFileName();
        this.type = transfer.getType();
        this.state = transfer.getState();
        this.size = getSizeString(transfer.getTotalBytes());
        this.nodeHandle = transfer.getNodeHandle() + "";
        this.path = getTransferPath(transfer);
        this.timeStamp = System.currentTimeMillis();
        this.error = getTranslatedErrorString(error);
        this.originalPath = transfer.getPath();
        this.parentHandle = transfer.getParentHandle();
    }

    public AndroidCompletedTransfer (SDTransfer transfer) {
        this.fileName = transfer.getName();
        this.type = MegaTransfer.TYPE_DOWNLOAD;
        this.state = MegaTransfer.STATE_COMPLETED;
        this.size = transfer.getSize();
        this.nodeHandle = transfer.getNodeHandle();
        this.path = removeLastFileSeparator(getSDCardTargetPath(transfer.getAppData()));
        this.timeStamp = System.currentTimeMillis();
        this.error = getString(R.string.api_ok);
        this.originalPath = transfer.getPath();
        this.parentHandle = INVALID_HANDLE;
        setIsOfflineFile(false);
    }

    public String getFileName() {
        return fileName;
    }

    public void setFileName(String fileName) {
        this.fileName = fileName;
    }

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }

    public int getState() {
        return state;
    }

    public void setState(int state) {
        this.state = state;
    }

    public String getSize() {
        return size;
    }

    public void setSize(String size) {
        this.size = size;
    }

    public String getNodeHandle() {
        return nodeHandle;
    }

    public void setNodeHandle(String nodeHandle) {
        this.nodeHandle = nodeHandle;
    }

    public String getPath() {
        return path;
    }

    public void setPath(String path) {
        this.path = path;
    }

    public boolean getIsOfflineFile() {
        return isOfflineFile;
    }

    public void setIsOfflineFile(boolean isOfflineFile) {
        this.isOfflineFile = isOfflineFile;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public long getTimeStamp() {
        return timeStamp;
    }

    public void setTimeStamp(long timeStamp) {
        this.timeStamp = timeStamp;
    }

    public String getError() {
        return error;
    }

    public void setError(String error) {
        this.error = error;
    }

    public String getOriginalPath() {
        return originalPath;
    }

    public void setOriginalPath(String originalPath) {
        this.originalPath = originalPath;
    }

    public long getParentHandle() {
        return parentHandle;
    }

    public void setParentHandle(long parentHandle) {
        this.parentHandle = parentHandle;
    }

    /**
     * Remove the last character of the path if it is a file separator.
     *
     * @param path  path of a file.
     * @return The path without the last item if it is a file separator.
     */
    private String removeLastFileSeparator(String path) {
        if (!isTextEmpty(path) && path.charAt(path.length() - 1) == '/') {
            path = path.substring(0, path.length() - 1);
        }

        return path;
    }

    /**
     * Gets the path of a transfer.
     *
     * @param transfer  MegaTransfer from which the path has to be obtained
     * @return  The path of the transfer.
     */
    private String getTransferPath(MegaTransfer transfer) {
        MegaApplication app = MegaApplication.getInstance();

        switch (type) {
            case MegaTransfer.TYPE_UPLOAD:
                setIsOfflineFile(false);
                MegaNode parentNode = app.getMegaApi().getNodeByHandle(transfer.getParentHandle());

                if (parentNode != null) {
                    return removeLastFileSeparator(getParentFolderPath(parentNode));
                }
                break;

            case MegaTransfer.TYPE_DOWNLOAD:
                String path = transfer.getParentPath();
                File offlineFolder = getOfflineFolder(app, OFFLINE_DIR);
                setIsOfflineFile(!isTextEmpty(path) && offlineFolder != null && path.startsWith(offlineFolder.getAbsolutePath()));

                if (isOfflineFile) {
                    path = removeInitialOfflinePath(path, transfer.getNodeHandle());
                }
                return removeLastFileSeparator(path);
        }

        setIsOfflineFile(false);
        return "";
    }

    protected AndroidCompletedTransfer(Parcel in) {
        id = in.readLong();
        fileName = in.readString();
        type = in.readInt();
        state = in.readInt();
        size = in.readString();
        nodeHandle = in.readString();
        path = in.readString();
        isOfflineFile = in.readByte() != 0;
        timeStamp = in.readLong();
        error = in.readString();
        originalPath = in.readString();
        parentHandle = in.readLong();
    }

    public static final Creator<AndroidCompletedTransfer> CREATOR = new Creator<AndroidCompletedTransfer>() {
        @Override
        public AndroidCompletedTransfer createFromParcel(Parcel in) {
            return new AndroidCompletedTransfer(in);
        }

        @Override
        public AndroidCompletedTransfer[] newArray(int size) {
            return new AndroidCompletedTransfer[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(id);
        dest.writeString(fileName);
        dest.writeInt(type);
        dest.writeInt(state);
        dest.writeString(size);
        dest.writeString(nodeHandle);
        dest.writeString(path);
        dest.writeByte((byte) (isOfflineFile ? 1 : 0));
        dest.writeLong(timeStamp);
        dest.writeString(error);
        dest.writeString(originalPath);
        dest.writeLong(parentHandle);
    }
}

File: app/src/main/java/mega/privacy/android/app/AndroidCompletedTransfer.kt
package mega.privacy.android.app

import android.content.Context
import mega.privacy.android.app.MegaApplication.Companion.getInstance
import mega.privacy.android.app.objects.SDTransfer
import mega.privacy.android.app.presentation.extensions.getErrorStringId
import mega.privacy.android.app.utils.MegaNodeUtil.getNodeFolderPath
import mega.privacy.android.app.utils.OfflineUtils
import mega.privacy.android.app.utils.SDCardUtils
import mega.privacy.android.app.utils.StringResourcesUtils
import mega.privacy.android.app.utils.TextUtil
import mega.privacy.android.app.utils.Util
import mega.privacy.android.data.extensions.mapTransferType
import mega.privacy.android.domain.entity.transfer.Transfer
import mega.privacy.android.domain.entity.transfer.TransferState
import mega.privacy.android.domain.exception.MegaException
import nz.mega.sdk.MegaApiJava
import nz.mega.sdk.MegaError
import nz.mega.sdk.MegaTransfer

/**
 * The entity for completed transfer
 *
 * @property id
 * @property fileName
 * @property type
 * @property state
 * @property size
 * @property nodeHandle
 * @property path
 * @property isOfflineFile
 * @property timeStamp
 * @property error
 * @property originalPath
 * @property parentHandle
 */
class AndroidCompletedTransfer {
    var id: Long = 0
    val fileName: String?
    val type: Int
    val state: Int
    val size: String?
    val nodeHandle: String
    var path: String?
    var isOfflineFile = false
    val timeStamp: Long
    val error: String?
    val originalPath: String?
    val parentHandle: Long

    constructor(transfer: MegaTransfer, error: MegaError, context: Context) {
        fileName = transfer.fileName
        type = transfer.type
        state = transfer.state
        size = Util.getSizeString(transfer.totalBytes, context)
        nodeHandle = transfer.nodeHandle.toString()
        path = getTransferPath(transfer.parentPath, transfer.nodeHandle, transfer.parentHandle)
        timeStamp = System.currentTimeMillis()
        this.error = getErrorString(
            transfer.isForeignOverquota,
            error.errorCode,
            StringResourcesUtils.getTranslatedErrorString(error),
            context
        )
        originalPath = transfer.path
        parentHandle = transfer.parentHandle
    }

    constructor(transfer: Transfer, error: MegaException?, context: Context) {
        fileName = transfer.fileName
        type = transfer.transferType.mapTransferType()
        state = transfer.state.mapTransferState()
        size = Util.getSizeString(transfer.totalBytes, context)
        nodeHandle = transfer.nodeHandle.toString()
        path = getTransferPath(transfer.parentPath, transfer.nodeHandle, transfer.parentHandle)
        timeStamp = System.currentTimeMillis()
        this.error = error?.let {
            getErrorString(
                transfer.isForeignOverQuota,
                it.errorCode,
                context.getString(error.getErrorStringId()),
                context
            )
        }
        originalPath = transfer.localPath
        parentHandle = transfer.parentHandle
    }

    constructor(transfer: SDTransfer, context: Context) {
        fileName = transfer.name
        type = MegaTransfer.TYPE_DOWNLOAD
        state = MegaTransfer.STATE_COMPLETED
        size = transfer.size
        nodeHandle = transfer.nodeHandle
        path = removeLastFileSeparator(SDCardUtils.getSDCardTargetPath(transfer.appData))
        timeStamp = System.currentTimeMillis()
        error = context.getString(R.string.api_ok)
        originalPath = transfer.path
        parentHandle = MegaApiJava.INVALID_HANDLE
        isOfflineFile = false
    }

    /**
     * Remove the last character of the path if it is a file separator.
     *
     * @param path  path of a file.
     * @return The path without the last item if it is a file separator.
     */
    private fun removeLastFileSeparator(path: String?): String? =
        if (!TextUtil.isTextEmpty(path) && path?.last() == '/')
            path.substring(0, path.length - 1)
        else
            path

    /**
     * Gets the path of a transfer.
     *
     * @param parentPath The parent path related to this transfer.
     * @param nodeHandle Handle related to this transfer.
     * @param parentHandle Handle of the parent node related to this transfer.
     * @return  The path of the transfer.
     */
    private fun getTransferPath(parentPath: String, nodeHandle: Long, parentHandle: Long): String? {
        val app = getInstance()
        return when (type) {
            MegaTransfer.TYPE_UPLOAD -> {
                isOfflineFile = false
                removeLastFileSeparator(
                    getNodeFolderPath(app.megaApi.getNodeByHandle(parentHandle))
                )
            }

            MegaTransfer.TYPE_DOWNLOAD -> {
                var path = parentPath
                val offlineFolder = OfflineUtils.getOfflineFolder(app, OfflineUtils.OFFLINE_DIR)
                isOfflineFile =
                    !TextUtil.isTextEmpty(path) && offlineFolder != null && path.startsWith(
                        offlineFolder.absolutePath
                    )
                if (isOfflineFile) {
                    path = OfflineUtils.removeInitialOfflinePath(path, nodeHandle)
                }
                removeLastFileSeparator(path)
            }

            else -> {
                isOfflineFile = false
                null
            }
        }
    }

    /**
     * Gets the error string to show as cause of the failure.
     *
     * @param isForeignOverquota True if the transfer has failed with MEGAErrorTypeApiEOverquota
     *                           and the target is foreign, false otherwise.
     * @param errorCode Error of the transfer.
     * @param errorMessage Translated error message.
     * @param context [Context]
     * @return The error to show as cause of the failure.
     */
    private fun getErrorString(
        isForeignOverquota: Boolean,
        errorCode: Int,
        errorMessage: String,
        context: Context,
    ): String =
        if (errorCode == MegaError.API_EOVERQUOTA && isForeignOverquota)
            context.getString(R.string.error_share_owner_storage_quota)
        else
            errorMessage

    /**
     * Temporary mapper while transitioning completely from [MegaTransfer] to [Transfer].
     */
    private fun TransferState.mapTransferState(): Int = when (this) {
        TransferState.STATE_NONE -> MegaTransfer.STATE_NONE
        TransferState.STATE_QUEUED -> MegaTransfer.STATE_QUEUED
        TransferState.STATE_ACTIVE -> MegaTransfer.STATE_ACTIVE
        TransferState.STATE_PAUSED -> MegaTransfer.STATE_PAUSED
        TransferState.STATE_RETRYING -> MegaTransfer.STATE_RETRYING
        TransferState.STATE_COMPLETING -> MegaTransfer.STATE_COMPLETING
        TransferState.STATE_COMPLETED -> MegaTransfer.STATE_COMPLETED
        TransferState.STATE_CANCELLED -> MegaTransfer.STATE_CANCELLED
        TransferState.STATE_FAILED -> MegaTransfer.STATE_FAILED
    }
}
